AI/VCAI/Pathfinding/AIPathfinder.cpp

```diff
@@ -69,12 +78,29 @@ void AIPathfinder::updatePaths(std::vector<HeroPtr> heroes)
 		}
 
 		storageMap[hero] = nodeStorage;
-		nodeStorage->setHero(hero.get());
+		nodeStorage->setHero(hero, cb);
 
 		auto config = std::make_shared<AIPathfinding::AIPathfinderConfig>(cb, ai, nodeStorage);
 
-		logAi->debug("Recalculate paths for %s", hero->name);
-		cb->calculatePaths(config, hero.get());
+		calculationTasks.push_back(std::bind(calculatePaths, hero.get(), config));
+	}
+
+	int threadsCount = std::min(
+		boost::thread::hardware_concurrency(),
+		(uint32_t)calculationTasks.size());
+
+	if(threadsCount == 1)
```

Comment:
`boost::thread::hardware_concurrency()` could be zero in case of error, you should check that.

Reply:
How to handle 0? Retry or do not use parallelism (thread count = 1)

Reply:
For now I check if tread count is <= 1 and if it is I use synchronous processing

Reply:
Do not use.